framework/src/media/audio/audio_manager.c

```diff
@@ -140,121 +151,72 @@ static const struct audio_samprate_map_entry_s g_audio_samprate_entry[] = {
 
-		if ((type == 'c') && (card_id < CONFIG_AUDIO_MAX_INPUT_CARD_NUM) && (device_id < MAX_AUDIO_DEVICE_NUM)) {
-			snprintf(g_audio_in_cards[card_id].card_path, strlen("/dev/audio/") + strlen(dir_entry->d_name) + 1, "%s%s",
-					 "/dev/audio/", dir_entry->d_name);
-
-			g_actual_audio_in_card_id = card_id;
-			g_audio_in_cards[card_id].status = AUDIO_CARD_READY;
-			g_audio_in_cards[card_id].device_id = device_id;
-			g_audio_in_card_num++;
-
-			ret = get_supported_sample_rate(INPUT, g_audio_in_cards[card_id].card_path);
-			ASSERT(ret == AUDIO_MANAGER_SUCCESS);
-
-			medvdbg("Found an input audio card, id=%d, count=%d,  card_path=%s\n", card_id, g_audio_in_card_num,
-					g_audio_in_cards[card_id].card_path);
+		if (card_type == INPUT) {
+			max_card_num = CONFIG_AUDIO_MAX_INPUT_CARD_NUM;
+			type_chr = 'c';
+			card = g_audio_in_cards;
 		}
-	}
-
-	if (!g_audio_in_card_num) {
-		meddbg("Wrong g_audio_in_card_num\n");
-		ret = AUDIO_MANAGER_NO_AVAIL_CARD;
-	}
-
-#ifndef CONFIG_AUDIO_EXCLUDE_VOLUME
-	if (get_input_audio_volume() < 0) {
-		meddbg("Fail to get an input volume\n");
-		ret = AUDIO_MANAGER_DEVICE_FAIL;
-		goto error_out;
-	}
-	medvdbg("max_vol=%d, cur_vol=%d\n", g_audio_in_cards[g_actual_audio_in_card_id].config.max_volume,
-			g_audio_in_cards[g_actual_audio_in_card_id].config.volume);
-#endif
-
-error_out:
-	closedir(dir_info);
-	return ret;
-}
 
-audio_manager_result_t find_output_audio_card(void)
-{
-	int ret = AUDIO_MANAGER_SUCCESS;
-	uint8_t card_id = -1;
-	uint16_t device_id = -1;
-	char type;
-	struct dirent *dir_entry;
-	DIR *dir_info;
-
-	if (!(dir_info = opendir("/dev/audio/"))) {
-		return AUDIO_MANAGER_DEVICE_FAIL;
-	}
-
-	while ((dir_entry = readdir(dir_info)) != NULL) {
-		// TODO: Add cases for various drivers. Currently, identify 'pcm' drivers only.
-		if ((dir_entry->d_name[0] != 'p') || (dir_entry->d_name[1] != 'c') || (dir_entry->d_name[2] != 'm') ||
-			(sscanf(&dir_entry->d_name[3], "C%uD%u%c", &card_id, &device_id, &type) != 3)) {
+		/* Find next card if driver name isn't follow naming policy */
+		if (card_id >= max_card_num || device_id >= MAX_AUDIO_DEVICE_NUM) {
+			meddbg("Found wrong card card_id : %d device_id : %d\n", card_id, device_id);
 			continue;
 		}
 
-		if ((type == 'p') && (card_id < CONFIG_AUDIO_MAX_OUTPUT_CARD_NUM) && (device_id < MAX_AUDIO_DEVICE_NUM)) {
-			snprintf(g_audio_out_cards[card_id].card_path, strlen("/dev/audio/") + strlen(dir_entry->d_name) + 1,
-					 "%s%s", "/dev/audio/", dir_entry->d_name);
-
-			g_actual_audio_out_card_id = card_id;
-			g_audio_out_cards[card_id].status = AUDIO_CARD_READY;
-			g_audio_out_cards[card_id].device_id = device_id;
-			g_audio_out_card_num++;
-
-			ret = get_supported_sample_rate(OUTPUT, g_audio_out_cards[card_id].card_path);
-			ASSERT(ret == AUDIO_MANAGER_SUCCESS);
-
-			medvdbg("Found an output audio card, id=%d, count=%d,  card_path=%s\n", card_id, g_audio_out_card_num,
-					g_audio_out_cards[card_id].card_path);
+		if (type == type_chr) {
```

Comment:
I think this "if" can be omitted if line 178 is modified as my comment.

Reply:
same as above, I put explanation addtionally,
I modified this function to use both of INPUT & OUTPUT case.
Basically some values initialized for OUTPUT, but if param is INPUT, it changes some values.
so flow is below
1. change values
2. find direnty in audio driver directory
3. if driver name is related to card type, it set values.
4. after find all cards, it check driver founded or not.
5. if founded, decide actual card.

Reply:
Also please check it continue if card_id is greater than max_card_num(changed in line #178) or device id is grater than MAX_AUDIO_DEVICE_NUM this can not be added in if state.